### DISCUSSION TITLE
Better Task Scheduling for NX & DTE

### DISCUSSION THREAD

**Discussion by aramfe** [nrwl/nx#23433] [Feature Requests]
Hello everyone,

I'm reaching out with a feature idea that could significantly enhance our experience with NX and DTE. The core concept revolves around leveraging artificial intelligence (AI) to dynamically determine the most efficient sequence for executing tasks based on past runs. If that would be too complex, I would suggest using a manually set priority system. 

**AI-Driven Optimization:** Utilize machine learning algorithms to analyze previous task executions. By understanding patterns and dependencies, these algorithms could predict the optimal order for running tasks, thereby reducing overall execution time.

If too complex for now, an easy alternative would be a

**Priority-Based Targeting:** Tasks would be assigned a numerical value indicating their importance, with higher numbers signifying lower priority. This system would allow us to prioritize essential tasks that take longer, ensuring they're executed before the shorter ones.

Example: 

```
"nx": {
    "targets": {
      "build": {
        "inputs": [
          "default",
          "^production",
          "sharedGlobals"
        ],
        "outputs": [
          "{projectRoot}/dist"
        ],
        "dependsOn": [
          "^build"
        ],
        "cache": true, 
        "priority": 1
      }
    }
  }
```

The ultimate objective is to streamline the task execution process within DTE, aiming for the quickest possible completion times. By intelligently scheduling tasks, one can maximize productivity and efficiency without too much hassle. 

While I believe this feature has immense potential, I also recognize its complexity. Therefore, I suggest starting with a basic implementation focusing on priority-based targeting.

**Comment by StalkAltan** [comment]
Hi @aramfe, thanks for the feature request.

You may be surprised to hear that Nx Cloud already takes a variety of factors into account to keep execution as optimal as is reasonable, in a general case.

Some of the things we automatically prioritize:
1) Unblocking later work as quickly as possible to achieve greater parallelization potential
2) Running tasks with known longer durations earlier, so smaller tasks can "fill in the gaps" later in the execution
3) Running tasks that require retries up front to unblock their dependent work, when necessary

There is a lot of academic research on this topic, and the truth is that graph scheduling (DTE) is an np-complete problem. This means there is no computationally-feasible way to achieve the optimal solution for a given graph scheduling problem.

Here's an paper that goes over this, if you're interested: https://www.cs.montana.edu/bhz/classes/fall-2018/csci460/paper4.pdf

This is all to say, we do not have current plans to allow manual prioritization of tasks with DTE. It is far more likely for users to experience this kind of feature as a footgun that results in *longer* execution times, than improve their average runtimes with intimate knowledge of their graphs.
